### PR TITLE
Fix GPT-NeoX-20B past handling, attention computation

### PR DESCRIPTION
**Pull request by zphang**
# What does this PR do?

* Fixes GPT-NeoX-20B handing of the past object to correctly be used in .generate
* Swaps attention computation for one more similar in the original training code, to hopefully avoid NaNs
* Update docstring, removed unnecessary dropout configs in config object

Fixes # (issue)

https://github.com/huggingface/transformers/issues/17632 
https://github.com/huggingface/transformers/issues/17452 (Hopefully)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sgugger

### CODE REVIEW

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -635,7 +648,7 @@ def prepare_inputs_for_generation(self, input_ids, past=None, attention_mask=Non
             attention_mask = input_ids.new_ones(input_shape)
 
         # cut decoder_input_ids if past is used
-        if past is not None:
+        if past and past[0] is not None:
```
Comment (sgugger) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:651]:
```suggestion
        if past is not None and past[0] is not None:
```

**Review by sgugger — APPROVED** [review]
Thanks for the fix, I can confirm that with this PR, I get the same generations in float32 and float16 (whereas before, I get either a crappy one or some Nans in float16) for `EleutherAI/gpt-neox-20b`.

The cleaning up in the config LGTM, thanks for making the docstring on par with the defaults, and the two attributes you remove are not used anywhere.

**Comment by sgugger** [comment]
There are a few equivalence tests failing with the PR, if you can dive into it. Let us know if you need any help!

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 2]
```diff
@@ -635,7 +648,7 @@ def prepare_inputs_for_generation(self, input_ids, past=None, attention_mask=Non
             attention_mask = input_ids.new_ones(input_shape)
 
         # cut decoder_input_ids if past is used
-        if past is not None:
+        if past is not None and past[0] is not None:
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:651]: Out of curiosity why is the second statement needed here? The `past[0] is not None` part?

File: src/transformers/models/gpt_neox/configuration_gpt_neox.py [hunk 3]
```diff
@@ -38,32 +38,28 @@ class GPTNeoXConfig(PretrainedConfig):
 
 
     Args:
```
Comment (patrickvonplaten) [inline src/transformers/models/gpt_neox/configuration_gpt_neox.py:40]: Thanks for cleaning this up!

**Review by patrickvonplaten — APPROVED** [review]
Thanks for fixing @zphang !

**Comment by zphang** [comment]
I've run the tests locally and they pass, so I can't seem to reproduce the test errors. Can someone else give them a try?

**Comment by sgugger** [comment]
The tests pass on GPU but not on CPU on my side. So doing
```
CUDA_VISIBLE_DEVICES="" pytest tests/models/gpt_neox/test_modeling_gpt_neox.py
```
reproduces the failure.

**Comment by sgugger** [comment]
Thanks again! Nice to be able to use GPT-Neo-X in float16 for generations :-)
